docs/design-notes/reverse-reads.md

```diff
@@ -0,0 +1,121 @@
+
+    # Reverse read
+    SELECT * FROM mytable WHERE pk = 1 ORDER BY ck DESC;
+
+If the table's native clustering order is DESC, then a read with ASC
+order is considered reverse.
+
+## Legacy format
+
+The legacy format is how scylla handled reverse queries internally. We
+are in the process of migrating to the native reverse format, but for
+now coordinator-side code still uses the legacy format.
+
+### Request
+
+The `query::partition_slice::options::reversed` flag is set.
+Clustering ranges in both `query::partition_slice::_row_ranges` and
+`query::specific_ranges::_ranges`
+(`query::partition_slice::_specific_ranges`) are half-reversed: they
+are ordered in reverse, their end bound used as start bound and
+vice-versa.
+Example:
+
+For the clustering keys (ASC order): `ck1`, `ck2`, `ck3`, `ck4`, `ck5`,
+`ck6`.
+A `_row_ranges` field of a slice might contain this:
+
+    [ck1, ck2], [ck4, ck5]
+
+The legacy reversed version would look like this:
+
+    [ck4, ck5], [ck1, ck2]
+
+Note how the ranges themselves are the same (bounds not reversed), it is
+just the range vector itself that is reversed.
+
+### Result
+
+Results are ordered with the reversed clustering order with the caveat
+that range-tombstones are ordered by their end bound, using the native
+schema's comparators. For example given the following partition:
+
+    ps{pk1}, sr{}, cr{ck1}, rt{[ck2, ck4)}, cr{ck2}, cr{ck3}, cr{ck4}, ck{ck5}, pe{}
+
+The legacy reverse format equivalent of this looks like the following:
+
+    ps{pk1}, sr{}, cr{ck5}, rt{[ck2, ck4)}, cr{ck4}, cr{ck3}, cr{ck2}, ck{ck1}, pe{}
+
+Note:
+* Only clustering elements change;
+* Range tombstone's bounds are not reversed;
+* Range tombstones can be ordered off-by-one due to native schema
+  comparators used: `rt{[ck2, ck4)}` should be ordered *after*
+  `cr{ck4}`.
+
+Legend:
+* ps = partitions-tart
+* sr = static-row
+* cr = clustering-row
+* rt = range-tombstone
+* pe = partition-end
+
+## Native format
+
+The native format uses ordering equivalent to that of a table with
+reverse clustering format. Using `mytable` as an example, the native
+reverse format would be an identical table `my_reverse_table`, which
+uses `CLUSTERING ORDER BY (ck DESC);`. This allows middle layers in a
+read pipeline to just use a schema with reversed clustering order and
+process the reverse stream as normal.
+
+### Request
+
+The `query::partition_slice::options::reversed` flag is set as in the
+legacy format. Clustering ranges in both
+`query::partition_slice::_row_ranges` and
+`query::specific_ranges::_ranges`
+(`query::partition_slice::_specific_ranges`) are fully-reversed: they
+are ordered in reverse, their bound being swapped as well.
+Example:
+
+For the clustering keys (ASC order): `ck1`, `ck2`, `ck3`, `ck4`, `ck5`,
+`ck6`.
+A `_row_ranges` field of a slice might contain this:
+
+    [ck1, ck2], [ck4, ck5]
+
+The native reversed version would look like this:
+
+    [ck5, ck4], [ck2, ck1]
```

Comment:
But partition_slice is in idl, and we're not changing the wire format.

Reply:
This change would be binary-compatible on the wire. But we are not changing it anyway, this only happens on the replica.

Reply:
How would it be binary compatible? Different bits are sent, even if the IDL definition is unchanged. It's the worst case of compatibility because it appears compatible but isn't.

Please add comments to the IDL and to partition_slice to note the difference. We expect that IDL and in-memory representation be the same, but here they aren't, this can be a major surprise to someone reading the code. It would be ideal to introduce a new name (and have IDL confusing_partition_slice, C++ partition_slice and C++ query_ordered_partition_slice) but it's a huge amount of churn so let's be satisfied with comments.

Reply:
Done